query/src/main/java/org/infinispan/query/backend/QueryInterceptor.java

```diff
@@ -145,78 +147,77 @@ private Object handleDataWriteCommand(InvocationContext ctx, DataWriteCommand co
-               // https://issues.redhat.com/browse/ISPN-11731
-               return asyncValue(processChange(rCtx, cmd, cmd.getKey(), oldValue, entry2.getValue())
-                     .thenApply(ignore -> rv));
+         boolean unreliablePrevious = unreliablePreviousValue(command);
+         if (ctx.isInTxScope()) {
+            Map<Object, Object> oldValues = getOldValuesMap((TxInvocationContext<?>) ctx);
+            for (Object key : command.getAffectedKeys()) {
+               registerOldValue(ctx, key, unreliablePrevious, oldValues);
             }
             return rv;
-         });
+         } else {
+            return asyncValue(allOf(cmd.getAffectedKeys().stream().map(key -> indexIfNeeded(rCtx, cmd, unreliablePrevious, key)).toArray(CompletableFuture[]::new)));
+         }
+      });
+   }
+
+   private void registerOldValue(InvocationContext ctx, Object key, boolean unreliablePrevious, Map<Object, Object> oldValues) {
+      CacheEntry<?, ?> entryTx = ctx.lookupEntry(key);
+      if (entryTx != null && (entryTx.getValue() != null || !unreliablePrevious)) {
+         ReadCommittedEntry<?, ?> mvccEntry = (ReadCommittedEntry<?, ?>) entryTx;
+         oldValues.putIfAbsent(key, mvccEntry.getOldValue());
       }
    }
 
-   private Map<Object, Object> registerOldValues(TxInvocationContext ctx) {
+   private Map<Object, Object> getOldValuesMap(TxInvocationContext<?> ctx) {
       return txOldValues.computeIfAbsent(ctx.getGlobalTransaction(), gid -> {
          ctx.getCacheTransaction().addListener(() -> txOldValues.remove(gid));
          return new HashMap<>();
       });
    }
 
-   private Object handleManyWriteCommand(InvocationContext ctx, WriteCommand command) {
-      if (command.hasAnyFlag(FlagBitSets.SKIP_INDEXING)) {
-         return invokeNext(ctx, command);
+   private CompletableFuture<?> indexIfNeeded(InvocationContext rCtx, WriteCommand cmd, boolean unreliablePrevious, Object key) {
+      CacheEntry<?, ?> entry = rCtx.lookupEntry(key);
+      boolean isStale = false;
+      Object old = null;
+      if (entry instanceof MVCCEntry) {
+         ReadCommittedEntry<?, ?> mvccEntry = (ReadCommittedEntry<?, ?>) entry;
+         isStale = !mvccEntry.isCommitted();
+         old = unreliablePrevious ? UNKNOWN : mvccEntry.getOldValue();
       }
-      if (ctx.isInTxScope()) {
-         Map<Object, Object> oldValues = registerOldValues((TxInvocationContext) ctx);
-         for (Object key : command.getAffectedKeys()) {
-            CacheEntry entry = ctx.lookupEntry(key);
-            if (entry != null && !entry.isChanged() && (entry.getValue() != null || !unreliablePreviousValue(command))) {
-               oldValues.putIfAbsent(key, entry.getValue());
-            }
+      if (entry != null && entry.isChanged() && !isStale) {
+         if (log.isDebugEnabled()) {
+            log.debugf("Try indexing command '%s',key='%s', oldValue='%s', stale='%s'", cmd, key, old, isStale);
```

Comment:
`isStale` is always `false`

Reply:
```isStale``` can be ```true``` if the entry is not committed:

```
mvn -Dtest=IndexingDuringStateTransferTest test
grep -R "stale='true'" target/infinispan-infinispan-query.log
```

```
10:11:30,147 DEBUG (jgroups-7,IndexingDuringStateTransferTest-NodeA:[]) [QueryInterceptor] Skipping indexing for command 'PutKeyValueCommand{key=k, value=AnotherGrassEater{name='Fluffy', blurb='Very cute.'}, flags=[CACHE_MODE_LOCAL, SKIP_LOCKING, SKIP_REMOTE_LOOKUP, PUT_FOR_STATE_TRANSFER, SKIP_SHARED_CACHE_STORE, SKIP_OWNERSHIP_CHECK, IGNORE_RETURN_VALUES, SKIP_XSITE_BACKUP, IRAC_STATE], commandInvocationId=CommandInvocation:IndexingDuringStateTransferTest-NodeA:148, putIfAbsent=false, valueMatcher=MATCH_ALWAYS, metadata=InternalMetadataImpl{actual=EmbeddedMetadata{version=null}, created=-1, lastUsed=-1}, internalMetadata=null, successful=true, topologyId=65}',key='k', oldValue='AnotherGrassEater{name='Fluffy', blurb='Very cute.'}', stale='true'
```

Reply:
I mean `isStale` is always `false` in this log message :) 
```suggestion
            log.debugf("Try indexing command '%s',key='%s', oldValue='%s', stale='false'", cmd, key, old);
```

Reply:
no prob!

Reply:
fixed @pruivo